### PR TITLE
Refactor `progress` into an evented model

### LINKED ISSUE #3032: [SPIKE] Investigate future of progress bar architecture
Currently our `progress` objects inherit from `tqdm` allowing them to behave like terminal progress bars when there is no active viewer, and be added to the viewer when there is one. See [this discussion](https://github.com/sofroniewn/waver/pull/5#discussion_r666692376) for how useful that can be.

The ability to "alias" our own progress objects as `tqdm` can be very useful for code which already has terminal progress bars available, and @Carreau initially suggested the possibility of a patch we could use for code that we suspect might have a `tqdm` progress bar:

```python
@contextmanager
def patch()
     import tqdm
     orig = tqdm.tqdm
     tqdm.tqdm = progress # the progress you define.
     yield 
     tqdm.tqdm = orig
```

which could be used like:

```python
def action():
      with patch():
        dask.do_stuff() # maybe use tqdm
button.click(action)
```

In an ideal scenario, users would be able to get progress bars in the napari viewer without making any changes to their code - potentially there is an avenue for a patch similar to the above that is even less invasive e.g. it patches the update method of any progress bars used to emit an event which we then connect to in the viewer (@jni originally suggested something like this). 

This issue is meant to track suggestions, discussions and decisions for the desired long term behaviour and architecture of our progress bars, and how we can provide maximal flexibility with minimal changes in user code.

### PR DESCRIPTION
**Pull request by DragaDoncila**
# Description
Following various discussions about progress bar architecture and concerns with its tight coupling with `Qt` this PR refactors `progress` into an `Evented` model based on an idea of @tlambert03's.

Rather than `progress` storing a `QtProgressBar` widget to perform the necessary updates, `progress` now emits events for each update that might require a change in the `Viewer`. The `ActivityDialog` is then connected to `progress` via a new `progress` class attribute - `all_progress` - an `EventedSet` that lets the `ActivityDialog` know when new progress bars need to be created, or old ones destroyed. This pattern is much closer to our usual model behaviour, and completely decouples the `progress` objects from their `Qt` view.

This is still a WIP but I wanted to get early feedback on whether we like this direction. @tlambert03 and @nclack are generally on board, but would love for @sofroniewn and @jni to have a look as well. As we wrap up design work on progress bars, it would be nice to have some certainty about the backend architecture before embarking on redesigning them.

Still outstanding to implement/decide/discuss:
- There's still some output to the terminal from `tqdm` which needs to be suppressed when we have a gui
- Multiple viewers seem to be handled just fine, with the ProgressBars being added to the appropriate viewer - I find this kinda weird and don't really get why it's working since I thought `ActivityDialog`s would share the same `progress.all_progress` instance?
- We need a way to close `progress` objects if the `Viewer` is closed
- We need to decide what happens with any existing `progress` objects when a `Viewer` is launched - do we add them to the viewer? Leave them alone? Can this even happen?

## Type of change
Refactor of existing functionality.

# References
Closes #3032.

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

File: napari/_qt/widgets/qt_progress_bar.py
Comment (sofroniewn) [inline napari/_qt/widgets/qt_progress_bar.py]:
```suggestion
class QtLabeledProgressBar(QWidget):
```

As this object inherits from a `QWidget`, maybe it should be named `Qt...`

File: napari/_qt/widgets/qt_progress_bar.py
Comment (sofroniewn) [inline napari/_qt/widgets/qt_progress_bar.py]:
```suggestion
        self.progress = progress
```

If this is a `Progress` instance

File: napari/_qt/widgets/qt_progress_bar.py
Comment (sofroniewn) [inline napari/_qt/widgets/qt_progress_bar.py]:
```suggestion
        self.qt_progress_bar = QProgressBar()
```
I don't feel so strongly about this one, but could be considered - could also be left

File: napari/utils/progress.py
Comment (sofroniewn) [inline napari/utils/progress.py]: I maybe don't quite get this. `progress` itself in some sense corresponds to one progress bar right? why is it that then that one progress bar should keep track of all the other progress bars?

i.e. if i make three progress bars, will each of them have an `all_progress` attribute that is the same (or different?) evented set that contains all three?

maybe this should just be a new object that isn't an attribute here. i guess the problem with that approach is that there might be no way to connect to register to it in the first place, unless maybe that object is like a global registry of progress bars that is accessed in the progress init

**Review by sofroniewn — COMMENTED** [review]
In general i love the direction. I left some minor comments about renaming.

My only major question is about the `all_progress` attribute. I think I see why it is needed, but maybe that can be explained a bit more just so i make sure i get it. Explanation should ultimately go in comment or doc string too, as i see how it has potential to be confusing

One additional thought is I'm curious now actually if i wanted to list all the active progress bars in the API how would i do that, and then how would one do that after this PR merges.

All in all though, very nice direction

File: napari/utils/progress.py
Comment (tlambert03) [reply napari/utils/progress.py]: > why is it that then that one progress bar should keep track of all the other progress bars?

It's a class attribute, not an instance attribute.  Note how "self" is added to progress.all_progress (not self.all_progress).  I think a class attribute is preferable to an independent object to track instances of progress (and this is a common pattern, see how dask tracks Callback instance for example).

Comment (nclack) [reply napari/utils/progress.py]: Adding docs around this seems important.  I think this is a nice design, but it needs to be called out.  The fact that the progress() constructor has a side-effect - it's added to a singleton, evented collection - needs to be pointed out in the constructor docs.

Comment (sofroniewn) [reply napari/utils/progress.py]: ahhhh ok - totally missed that, but makes total sense. Class attribute sounds like a good way to go. I like it.

Unrelated to this PR, but do we want `viewer.all_viewers` where `all_viewers` is a class attribute and an evented set. I know in the past i've resisted that idea, but maybe that's a nice pattern to use there too. Can discuss elsewhere, but wanted to flag now

Comment (tlambert03) [reply napari/utils/progress.py]: lol... yes this is exactly where that came up before.  we wanted to make `current()` a classmethod on `Viewer`, but you found it odd that you could do `viewer.current()` on an instance and not always get back the same instance on which it was called.  So we tucked it away on _QMainWindow, ... and thats how napari.current_viewer works.  (i.e.  the equivalent of `viewer.all_viewers` is [_QtMainWindow.__instances](https://github.com/napari/napari/blob/315e2bf256f2e70fe96202beadeb8fb1458aca2b/napari/_qt/qt_main_window.py#L71))

I'd be supportive of adding it to Viewer... would be much cleaner.

Comment (sofroniewn) [reply napari/utils/progress.py]: > we wanted to make current() a classmethod on Viewer, but you found it odd that you could do viewer.current() on an instance and not always get back the same instance on which it was called.

Indeed, indeed. Potentially irrationally, I still don't like `viewer.current()`. However, `viewer.all_viewers.current()` makes much more sense to me. I'm certainly not trying to be pedantic here, but publicly exposing the set of all viewers on a viewer, and then allowing users to find the current viewer through that set of all viewers makes a lot more sense to me then straight up `viewer.current`.

What do you think? Is @jni on board with `viewer.all_viewers` and `viewer.all_viewers.current()`?

Comment (DragaDoncila) [reply napari/utils/progress.py]: Sorry, it's been a while on this one but picking back up and hoping to have ready for review by Monday. 

For what it's worth when class attributes were first introduced to me (in Java `static` class attributes which is not *quite* the same) they were introduced as a way to either track information common to all instances of the class (we commonly do this with class constants - just class attributes in all caps) or to track "meta" information about the class e.g. how many instances have been created if you're trying to keep some sort of unique ID for each one.

So to me `Viewer.current()` or `Viewer.all_viewers.current()` seems perfectly natural, with particular emphasis on `Viewer` not `viewer` because we're talking about the class itself not an instance.

Comment (tlambert03) [reply napari/utils/progress.py]: I like `Viewer.current()` ... but I don't like `Viewer.all_viewers.current`

> with particular emphasis on `Viewer` not `viewer` because we're talking about the class itself not an instance.

I agree ... this is the way to think about it... but it has to be acknowledged that the method _will_ also appear on the instance, and I think that's @sofroniewn's hesitation

Comment (DragaDoncila) [reply napari/utils/progress.py]: >  that the method will also appear on the instance, and I think that's @sofroniewn's hesitation

yeah that is a key difference with Python, and it's caught me out before when I was teaching class attributes lol. As far as I know there's no way to have a class attribute not accessible through the instance right? Can we write an accessor method  that warns if you try to access it from the instance? Or even just raises an error?

Comment (tlambert03) [reply napari/utils/progress.py]: > Can we write an accessor method that warns if you try to access it from the instance? Or even just raises an error?

you _could_ ... like this

```py
class ClassOnlyAttribute:
    def __init__(self, value):
        self.value = value

    def __set_name__(self, obj, name):
        self.name = name

    def __get__(self, instance, owner) -> None:
        if instance is None:
            return self.value
        raise AttributeError(
            f"{self.name!r} attribute of {owner.__name__!r} is class-only"
        )


class Viewer:
    current = ClassOnlyAttribute([])

```

but my personal preference would be _not_ to.  Whether it surprises someone at first or not, this _is_ how python works (class attributes are accessible on instance) ... and I think we need a bit of a better reason to override that (since then you risk surprising someone who _does_ know how python works and I'd argue that's equally bad). That is, I'm still looking for a more compelling _problem_ with the fact that, theoretically, someone could use `viewer.current` as well as `Viewer.current`.  So what?

Comment (DragaDoncila) [reply napari/utils/progress.py]: True, I guess the usual concern here is that you mutate the instance attribute and then it becomes out of sync with the class attribute, but if a user is trying to mutate `viewer.current()` they're dancing to their own tune anyway :woman_dancing:

Comment (sofroniewn) [reply napari/utils/progress.py]: > I like Viewer.current() ... but I don't like Viewer.all_viewers.current

@tlambert03 does that extend to viewer.all_viewers.current? What don't you like about it?

> but it has to be acknowledged that the method will also appear on the instance, and I think that's @sofroniewn's hesitation

If i'm the only one who feels this way I can def be overruled!! What does @nclack think about all this? @jni preferences?

Comment (tlambert03) [reply napari/utils/progress.py]: > @tlambert03 does that extend to viewer.all_viewers.current? What don't you like about it?

Yeah, they're one and the same (make one you get the other).  I don't like it because I think `Viewer.current()` is very idiomatic python... and putting an additional all_viewers thing in there feels unnecessary (unless we specifically want to expose the full, list, which hasn't been a part of the discussion and I don't  see the need at the moment). 

Basically, I feel like we're finding a work around for a non-issue

Comment (sofroniewn) [reply napari/utils/progress.py]: > Yeah, they're one and the same (make one you get the other). 

yup ok, that's what i thought

> unless we specifically want to expose the full, list, which hasn't been a part of the discussion and I don't see the need at the moment

The full list seems useful to me, I could imagine cases where you might want to iterate overall the viewers.

> Basically, I feel like we're finding a work around for a non-issue

Ok I still find it much cleaner to have the extra list in there, and much less confusing to think about getting the current from that list. But I will defer to the majority here!!

Comment (nclack) [reply napari/utils/progress.py]: I think `Viewer.current()` works. It is brief and reasonably easy to remember.  The Viewer class might be the most discoverable location for it.  A similar pattern is used in the html dom i.e. [window.getCurrent()](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/windows/getCurrent).

I'd be ok with other syntax's too. 

One of the assumptions underlying all this is that we have one global application and that its state is spread out among these various classes.  We could think about the value of changing that - encapsulating some of this stuff in an Application object.  That's outside the scope of this PR though (imo).

Comment (sofroniewn) [reply napari/utils/progress.py]: > I think Viewer.current() works. It is brief and reasonably easy to remember. The Viewer class might be the most discoverable location for it. A similar pattern is used in the html dom i.e. window.getCurrent().

ok i can live with that

> We could think about whether the value of changing that - encapsulating some of this stuff in an Application object. That's outside the scope of this PR though (imo).

i love that, but agreed - let's take discussion elsewhere/ start a new issue. i'm not sure if there is a good one

**Comment by DragaDoncila** [comment]
I think this is now ready for a thorough review @tlambert03.

I've addressed most of the issues that were outstanding, but would love some feedback on a couple of things in particular:

### Printing in terminal vs. running in GUI
Since `progress` no longer has any knowledge of the GUI, we set `self.gui=True` when the `QtActivityDialog` gets notified about a `progress` object being created. However, `super().__init__` calls `refresh`, so we were still getting an initial print in the terminal. `tqdm` has a `disable` attribute you can pass in at construction, but that also disables the creation of many attributes we use like the description. 

So, to temporarily disable printing to the terminal *before* we have a chance to tell `progress` it has a gui, I have a very janky `is_init` attribute which I check in the `display` method before calling `super().display`. This is not perfect because that means even when we are in headless, we don't get the initial progress bar printed, until its first update. Any ideas here would be appreciated.

### Threading and `progress.all_progress`
In the `threading_progress` tests we make a few workers but don't necessarily start them. This means `progress` objects are added to `all_progress`, but since the workers never finish, they're not removed. Is there any danger to them living in there for the remainder of the test suite? Should I be addressing this somehow e.g. deleting the worker at the end of the test? And should we have a catchall of some sort to clean up workers that may not have been started - if I run a script and don't start a worker created there and the script finishes, what is the lifetime of these `progress` objects?

### Behaviour on `viewer.close()`
If the user closes the `viewer` while a `progress` object is still iterating, we `close` that object by connecting to `activity_dialog.destroyed`. This seems to work without any issues, but I've noticed that `tqdm`'s behaviour is to [still `yield` out the rest of the iterable](https://github.com/tqdm/tqdm/blob/fc69d5dcf578f7c7986fa76841a6b793f813df35/tqdm/std.py#L1167). This is existing behaviour, and is identical to what was happening before this refactor, but I'm not sure if it's what we want... could this lead to trouble in any way?

File: napari/utils/progress.py
Comment (sofroniewn) [inline napari/utils/progress.py:11]: should this also inherit from EventedModel so the `EmitterGroup` doesn't need to be manually added? i think this can come in a followup PR, but just asking now

**Review by sofroniewn — APPROVED** [review]
Looks great! I left one question, but I don't think it should hold up merge. Maybe @tlambert03 can take another look before merge, especially at the questions above, but this is good to go from my perspective

File: napari/_qt/widgets/qt_progress_bar.py
Comment (tlambert03) [inline napari/_qt/widgets/qt_progress_bar.py]: type hints are always helpful... makes it easier to immediately recognize, "ok this takes a ____".  (otherwise reader has to go search the code for where `QtLabeledProgressBar()` is used to see what the object is)

File: napari/_qt/dialogs/qt_activity_dialog.py
Comment (tlambert03) [inline napari/_qt/dialogs/qt_activity_dialog.py:117]: ideally they would be registered?
It's not dissimilar to a layerlist that already has layers in it right?  If we created a new viewer window for a viewer model that already had layers, we'd want them to show up too

File: napari/_qt/dialogs/qt_activity_dialog.py
Comment (tlambert03) [inline napari/_qt/dialogs/qt_activity_dialog.py]:
```suggestion
        for prog in event.removed:
            self.close_progress_bar(prog)
        for prog in event.added:
            self.make_new_pbar(prog)
```

File: napari/utils/progress.py
Comment (tlambert03) [inline napari/utils/progress.py]: slight preference for not repeating the word `progress` here ... `progress.instances`?  `progress.all`?
also, give it a type hint: `EventedSet[progress]`

**Review by tlambert03 — APPROVED** [review]
haven't tested this locally but it looks like a good direction!  couple minor comments

File: napari/_qt/dialogs/qt_activity_dialog.py
Comment (DragaDoncila) [reply napari/_qt/dialogs/qt_activity_dialog.py:117]: @tlambert03 yeah I thought about that too, but I didn't know if it was the best user experience to lift progress bars started in the terminal into the viewer. I can see preferences for it and against it :shrug:

File: napari/utils/progress.py
Comment (DragaDoncila) [reply napari/utils/progress.py]: I've gone with `_instances` which is also what "native" `tqdm` uses

File: napari/utils/progress.py
Comment (DragaDoncila) [reply napari/utils/progress.py:11]: hmm I'm not sure. I don't know much about `EventedModel`s but I see them used mainly when we want to connect events to attributes of classes. While `value` and `description` are directly tied to `progress` attributes, `overflow` and `eta` aren't, so I'm not sure what the benefits of inheriting from `EventedModel` would be.

Comment (sofroniewn) [reply napari/utils/progress.py:11]: Yeah, maybe it's more appropriate for something with more properties, was just a thought. we can revisit in future if we want

**Review by sofroniewn — COMMENTED** [review]
I think this is now good to go, all tests passing! I'll merge tomorrow unless hearing anything more. Thanks!!

**Pull request merged**
